File: src/sort/DSU.java
package sort;



/*并查集模板*/

/**
 * @author 马世臣
 * @// TODO: 2020/9/2  */
public class DSU {

    private int[] parent;
    public DSU(int n){
        parent=new int[n];
        for (int i=0;i<n;i++){
            parent[i]=i;
        }
    }

    public int find(int x){
        while (x!=parent[x]){
            parent[x]=parent[parent[x]];
            x=parent[x];
        }
        return x;
    }

    public void Union(int x,int y){
        int px=find(x);
        int py=find(y);
        if(px<py){
            parent[py]=px;
        }else {
            parent[px]=py;
        }
    }

    public boolean isConnected(int x,int y) {
        return find(x)==find(y);
    }
}

File: src/Template/DSU.py
class DSU:
    def __init__(self, dataList):
        self.parent = {}
        self.size = {}
        self.setsCount = len(dataList)

        for x in dataList:
            self.parent[x] = x
            self.size[x] = 1

    def isExit(self, x):
        return x in self.parent

    def add(self, x):
        self.parent[x] = x
        self.size[x] = 1
        self.setsCount += 1

    def find(self, x):
        while x != self.parent[x]:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def union(self, x, y):
        px = self.find(x)
        py = self.find(y)
        if px != py:
            sx = self.size[px]
            sy = self.size[py]
            if sx < sy:
                px, py = py, px
            self.parent[py] = px
            self.size[px] += self.size[py]
            self.setsCount -= 1

    def isConnected(self, x, y):
        return self.find(x) == self.find(y)
